src/server.c

```diff
@@ -5654,6 +5658,7 @@ int prepareForShutdown(int flags) {
     {
         server.shutdown_mstime = server.mstime + server.shutdown_timeout * 1000;
         pauseClients(LLONG_MAX, CLIENT_PAUSE_WRITE);
+        sendGetackToReplicas();
```

Comment:
shoudln't the GETACK be eabove the pause?

Reply:
Does it matter? pauseClients only prevents incoming write commands, right?

Reply:
not really, it also prevents expiration, eviction, replica PINGs, and even hints modules that they must not do anything that results in replication.
maybe it doesn't affect a direct call to replicationFeedSlaves, but there might be an assert there some day, so i think it's a good idea to swap them.

come to think of it, what if we where already in a PAUSE?
maybe in that case we should avoid asking for ACK, but also what if the earlier PAUSE is removed earlier?
i suppose all it'll do is that the shutdown will be completed when we reach our timeout?
please look into this.

Reply:
OK, I'll skip GETACK if clients are already paused.

If clients are already paused, we still call `pauseClients(LLONG_MAX, CLIENT_PAUSE_WRITE)` here and unpause if and when shutdown fails, so if they were already paused before, the pause mode and timeout will be lost. I think we can accept this.

Reply:
so assuming we don't bother to re-pause or send a GETACK if already paused, two possible things can go wrong:
1. the other pause will expire before ours expires, which will mean:
  1. more traffic is forwarded to the replicas
  2. we don't have a chance to shutdown in reasonable time and wait till the timeout
  3. when the timeout is reached, we shudown and induce partial data loss
2. our shutdown timeout will expire first
  1. if we end up shutting down, that's fine
  2. if the shutdown fails, and we unpause, we could cause some damage to the original pause intent.

2.2 is solvable if we don't unpause what we didn't pause.
1.3 can be maybe solved if pause has a reference count, so the clients will only be unpaused if bother pausers ask to unpause (i don't like it)

WDTY?

Reply:
It's not good to let client pause timeout during shutdown. I don't think we should let that happen. Better set pause timeout to LONG_MAX so it can't timeout during shutdown.

What if clients are paused in mode ALL and we change it to WRITE? It will allow some requests during shutdown which would otherwise have been blocked. Perhaps it's better to leave keep ALL in that case and only set the timeout...?

What if clients are unpaused explicitly, for example if CLIENT UNPAUSE is called or CLUSTER FAILOVER is completed while shutdown is ongoing, which triggers unpauseClients, how bad is that? I think it's acceptable. If you think it's not acceptable, we can tweak unpauseClients so that it checks if shutdown is ongoing and refuse to unblock clients.

Reply:
ohh, i now notice that my previous message wasn't properly formatted, so it showed numbers from 1 to 7, instead of creating 2 levels like i meant to (edited).
but looking at your response, it looks like somehow you figured out what i meant?

anyway, this is becoming a little bit messy.
another way out (which i don't like either), is to introduce a separate pause flag or variable for each purpose, so that they don't override each other, and the pause is effective as long as either of them is set.

if we don't go that way, i think we need to make some compromises, and just decide which one is better..
obviously as you said, when we start out pause, we must not cause the pre-existing pause be less effective (like demoting an ALL, to WRITE).

but i'm a bit paranoid about cases that one of them will expire first and cause damage to the other.
maybe that shouldn't really happen, since it's unlikely that FAILOVER and SHUTDOWN will be triggered at the same time (user who does that might be misguided).

please propose a solution and analyze it and it's edge cases, and we'll see if we can agree on it.

Reply:
clientPause never demotes ALL to WRITE and never decreases a pre-existing timeout. It only extends them (ALL > WRITE > OFF).

```C
void pauseClients(mstime_t end, pause_type type) {
    if (type > server.client_pause_type) {
        server.client_pause_type = type;
    }

    if (end > server.client_pause_end_time) {
        server.client_pause_end_time = end;
    }
    ...
```

We can use this if we pause with an `end` timestamp matching the shutdown-timeout (instead of LONG_MAX) and don't explicitly unpause after a shutdown failure. Then, if clients were already paused for some other reason, they will remain paused after a shutdown failure.

The only downside of this -- that I can think of -- is that we need to add one cron period to the pause end-timestamp to make sure clients aren't unpaused before the cron function calls finishShutdown, and thus we might have to wait up to one cron period after a failed shutdown until clients are unpaused. I think it's acceptable.

This is what I had done originally and reverted in c941a29a2131267568b69221163c6bc54d13ae95.

WDYT?

Reply:
ok. i completely forgot that clientPause already has a mechanism for "conflict resolution".
so in that case it would have been best to just hand it with the request and let it solve it.
we do that with the pause_type, but not with the timeout, instead we try to keep manual control over it, and that can fail.

it's interesting that the FAILOVER command doing exactly the same (manual unpause), but maybe it's not as severe when there's just one player that does that, and it becomes more severe when there are two doing that trick.

We need to realize that the result of such a clash is partial data loss (doing an unpause, and then either shutting down or failing over, and losing writes)

Since we can't seem to find a good solution, maybe we need more input.
Please sum up the details and concerns of this issue as an open question in the top comment, and we'll call the core-team to review and respond.

one last note, i see FAILOVER uses LLONG_MAX, since mstime_t is long long, and we're using LONG_MAX, they're the same on 64bit, and i don't like to take any risk of different behavior in edge cases on 32 bit, so lets converge.

Reply:
In the past I've advocated for supporting a list of independent timeout events with their own level. I don't think the code will be all that complex, and it will much more clearly handle these edge cases.

Reply:
@zuiderkwast we discussed this in a core-team meeting, we would like to proceed by adding a separate pause flag per purpose so that pause is effective until the last one is cleared.
please look into that and see what you come up with.

Reply:
Per purpose... I can identify 3 purposes: SHUTDOWN, FAILOVER (cluster or standalone) and CLIENT PAUSE. Or do you have a different idea of purpose?

Currently, I think CLIENT UNPAUSE works when clients are paused during FAILOVER. We'll break that, so it's a breaking change. OK?

If CLIENT PAUSE is called twice with different timeouts and modes, the last one overwrites the previous ones, because it's the same purpose. OK?

Reply:
i think for CLIENT PAUSE we should keep the current logic of taking the longer pause and the stronger pause "level".
and since we have an explicit FAILOVER ABORT, i do think CLIENT UNPAUSE should not affect it (yes it's a breaking change).
And i guess we might wanna add SHUTDOWN ABORT.

Reply:
I added a commit "separate client pause/unpause per purpose" to this PR now. I hope it's acceptable. :grimacing: 

SHUTDOWN ABORT we do in a later PR?

Reply:
SHUTDOWN ABORT added in this PR, with tests.